### PR TITLE
Add new function `calculate_stats()` 

### PR DESCRIPTION
**Pull request by mrcaseb**
This implements #445

### CODE REVIEW

**Comment by mrcaseb** [comment]
It turned out that some stats are hard to compute using the playstats because some stats require information of other players, e.g. receiving_air_yards. 

The remaining stats probably need to be computed using pbp. I didn't start this yet

File: R/calculate_stats.R [hunk 1]
```diff
@@ -0,0 +1,263 @@
+calculate_stats <- function(seasons = nflreadr::most_recent_season(),
+                            summary_level = c("season", "week"),
+                            stat_type = c("player", "team")){
+
+  # testing
+  # seasons = 2023
+  # summary_level = "week"
+  # stat_type = "player"
+
```
Comment (guga31bb) [inline R/calculate_stats.R:9]: Do we want a default summary level (season?) and stat type (player?)

And error checking for valid summary level / stat type provided?

Comment (mrcaseb) [reply R/calculate_stats.R:9]: Both argument defaults and inputs checks are done in lines 10+11

**Comment by mrcaseb** [comment]
> What does this function do?

See here https://github.com/nflverse/nflfastR/blob/f59b2e672a3bb80b7ee570a63db0d070d0c46acb/NEWS.md?plain=1#L15

**Comment by jacobakaye** [comment]
I installed the development version, and the old function 'calculate_player_stats' still worked while I got the error with the new function. 

calculate_stats(pbp)
Error in calculate_stats(pbp) : could not find function "calculate_stats"

**Comment by mrcaseb** [comment]
@guga31bb and @tanho63 I have requested a review but honestly it's quite hard. Since it is a new function, it won't change anything at the moment. I think we can merge this and ask people to test it and look for possible bugs.

**Comment by mrcaseb** [comment]
- Fumble recovery stats now count all recoveries (not only defense). That's necessary to count fumble recoveries of offensive players, esp. when they result in a TD, e.g. Trey McBride 2024, wk2. cc: @alecglen 
- added new variable `fumble_recovery_tds` to account for the above
- Changed variable names of fumble recovery and penalty stats to make clear they don't belong to defense only
- documented the changes and improved variable explainer in `nfl_stats_variables`

**Comment by mrcaseb** [comment]
Changed variable names compared to the current stats approach are now documented here
https://github.com/nflverse/nflfastR/blob/cb83260032c33891c66fa2d33655337391ae6074/data-raw/nfl_stats_variables.R#L35-L65

This is the final call for review

**Comment by TheMathNinja** [comment]
> Changed variable names compared to the current stats approach are now documented here
> 
> https://github.com/nflverse/nflfastR/blob/cb83260032c33891c66fa2d33655337391ae6074/data-raw/nfl_stats_variables.R#L35-L65
> 
> This is the final call for review

Does this mean the new version is erasing the difference between solo tackles and tackles with assist? Or will it maintain the two categories and also offer a combined variable?

**Comment by mrcaseb** [comment]
> > Changed variable names compared to the current stats approach are now documented here
> 
> > 
> 
> > https://github.com/nflverse/nflfastR/blob/cb83260032c33891c66fa2d33655337391ae6074/data-raw/nfl_stats_variables.R#L35-L65
> 
> > 
> 
> > This is the final call for review
> 
> 
> 
> Does this mean the new version is erasing the difference between solo tackles and tackles with assist? Or will it maintain the two categories and also offer a combined variable?

It counts solo tackles, tackles with assist and tackle assists separately

**Comment by TheMathNinja** [comment]
> > > Changed variable names compared to the current stats approach are now documented here
> > 
> > 
> > > 
> > 
> > 
> > > https://github.com/nflverse/nflfastR/blob/cb83260032c33891c66fa2d33655337391ae6074/data-raw/nfl_stats_variables.R#L35-L65
> > 
> > 
> > > 
> > 
> > 
> > > This is the final call for review
> > 
> > 
> > Does this mean the new version is erasing the difference between solo tackles and tackles with assist? Or will it maintain the two categories and also offer a combined variable?
> 
> It counts solo tackles, tackles with assist and tackle assists separately

Got it!

**Comment by tanho63** [comment]
```r
new_stats <- calculate_stats(2023, summary_level = "week", stat_type = "player")

old_stats <- dplyr::bind_rows(nflreadr::load_player_stats(2023,stat_type = "offense"),nflreadr::load_player_stats(2023,stat_type = "defense"),nflreadr::load_player_stats(2023,stat_type = "kicking"))

setdiff(names(old_stats), names(new_stats))
#> [1] "recent_team"                   "interceptions"                 "sacks"                        
#> [4] "sack_yards"                    "dakota"                        "def_tackles"                  
#> [7] "def_fumble_recovery_own"       "def_fumble_recovery_yards_own" "def_fumble_recovery_opp"      
#>[10] "def_fumble_recovery_yards_opp" "def_safety"                    "def_penalty"                  
#>[13] "def_penalty_yards"      

setdiff(names(new_stats),names(old_stats))
#> [1] "passing_interceptions"     "sacks_suffered"            "sack_yards_lost"           "passing_cpoe"             
#> [5] "def_safeties"              "misc_yards"                "fumble_recovery_own"       "fumble_recovery_yards_own"
#> [9] "fumble_recovery_opp"       "fumble_recovery_yards_opp" "fumble_recovery_tds"       "penalties"                
#> [13] "penalty_yards"             "punt_returns"              "punt_return_yards"         "kickoff_returns"          
#> [17] "kickoff_return_yards"     
```

I see most things have equivalents, should we document these somewhere e.g. what new fields are avail + what fields are no longer avail?

File: tests/testthat/test-calculate_stats.R [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+  var_names <- nflfastR::nfl_stats_variables$variable
+
+  # Make sure variable names are listed in nflfastR::nfl_stats_variables$variable
+  expect_in(n1, var_names)
+  expect_in(n2, var_names)
+  expect_in(n3, var_names)
+  expect_in(n4, var_names)
+  expect_in(n5, var_names)
+
```
Comment (tanho63) [inline tests/testthat/test-calculate_stats.R:25]: we could snapshot the schema and type as well e.g. make sure skimr::skim() is consistent for field names and values? 

skim <- skimr::skim(new_stats) %>% 
    as.data.frame() %>% 
    dplyr::transmute(variable = skim_variable, type = skim_type, n_missing = round(n_missing, -1))

expect_snapshot_value(skim)

**Review by tanho63 — COMMENTED** [review]
helluva job tackling this! performance not _superb_ but I fully support dplyr readability here over data.table performance/maintenance tradeoff since nflreadr will pull precalculated values 99% of the time. a few questions about field differences, I think.

**Comment by mrcaseb** [comment]
> ```r
> new_stats <- calculate_stats(2023, summary_level = "week", stat_type = "player")
> 
> old_stats <- dplyr::bind_rows(nflreadr::load_player_stats(2023,stat_type = "offense"),nflreadr::load_player_stats(2023,stat_type = "defense"),nflreadr::load_player_stats(2023,stat_type = "kicking"))
> 
> setdiff(names(old_stats), names(new_stats))
> #> [1] "recent_team"                   "interceptions"                 "sacks"                        
> #> [4] "sack_yards"                    "dakota"                        "def_tackles"                  
> #> [7] "def_fumble_recovery_own"       "def_fumble_recovery_yards_own" "def_fumble_recovery_opp"      
> #>[10] "def_fumble_recovery_yards_opp" "def_safety"                    "def_penalty"                  
> #>[13] "def_penalty_yards"      
> 
> setdiff(names(new_stats),names(old_stats))
> #> [1] "passing_interceptions"     "sacks_suffered"            "sack_yards_lost"           "passing_cpoe"             
> #> [5] "def_safeties"              "misc_yards"                "fumble_recovery_own"       "fumble_recovery_yards_own"
> #> [9] "fumble_recovery_opp"       "fumble_recovery_yards_opp" "fumble_recovery_tds"       "penalties"                
> #> [13] "penalty_yards"             "punt_returns"              "punt_return_yards"         "kickoff_returns"          
> #> [17] "kickoff_return_yards"     
> ```
> 
> I see most things have equivalents, should we document these somewhere e.g. what new fields are avail + what fields are no longer avail? dakota and def_tackles for instance don't have direct equivs?

Done in the article that hosts the variables

File: tests/testthat/test-calculate_stats.R (hunk 2 above)
Comment (mrcaseb) [reply tests/testthat/test-calculate_stats.R:25]: done in e1f357fa0d55c96bb7944e5f1ed26edd0bee85d2

**Review by tanho63 — APPROVED** [review]

**Pull request merged**